### PR TITLE
Optional support for asciidoctor in dokka

### PR DESCRIPTION
**Pull request by mariotoffia**
Hi, I've tried to integrate Ascidoctor support in dokka with as less impact as I could. I'm *very* new to kotlin so please comment! 

Is it possible to accept this pull request as _experimental_ status? (did not find any annotation for it)

I've created a `AsciidocFormatService` that takes care of formatting text to _HTML_ via asciidoctor. I also created a `FormatDescriptor` for _html-as-asciidoc_. 

I've also created (of which I'm using) a `JavaAsciidocGenerator` that generates asciidoc formatted Javadoc. This impacted all of the Javadoc adapters since I've needed to provide with a `FormatService` in each as well as `TextTag`. This `FormatDescriptor` is set as _java-asciidoc_.

It supports asciidoctor markup as well as optional gems such as asciidoctor-diagram. It is a _compileOnly_ dependency and thus when using it in maven the following dependency must be set:
```xml
<plugin>
    <groupId>org.jetbrains.dokka</groupId>
    <artifactId>dokka-maven-plugin</artifactId>
    <version>${dokka.version}</version>
    <dependencies>
        <dependency>
            <groupId>org.asciidoctor</groupId>
            <artifactId>asciidoctorj</artifactId>
            <version>${version.asciidoctorj}</version>
        </dependency>
    </dependencies>
    <executions>
            <execution>
            <phase>pre-site</phase>
            <goals>
                <goal>dokka</goal>
                <goal>javadoc</goal>
            </goals>
        </execution>
    </executions>
    <configuration>
        <sourceLinks>
            <link>
                <dir>${project.basedir}/src/main/java</dir>
                <url>http://localhost-test</url>
                <urlSuffix>#L</urlSuffix>
            </link>
        </sourceLinks>
        <outputFormat>java-asciidoc</outputFormat>
        <jdkVersion>8</jdkVersion>
        <skip>false</skip>
        <additionalParams>
        --base-dir ${project.basedir}
        --attribute project_name="${project.name}"
        --attribute project_version=${project.version}
        --attribute project_desc="${project.description}"
        --attribute imagesdir="${asciidoctor.configuration.images-dir}"
        --attribute data-uri="${asciidoctor.configuration.images-dir}"
        --gem-path "${project.build.directory}/gems-provided"
        --require asciidoctor-diagram
        </additionalParams>
    </configuration>
</plugin>
```

In order to not change the configuration object I've only added additionalParams : String? = null. I need to pass several parameters to asciidoc in order for it to properly generate documentation; e.g. _--require_ etc.

This implementation certainly needs updates. Especially around the ugly hack in `DescriptorDocumentationParser` where I've hardcoded:

```kotlin
if (options.outputFormat == "html-as-asciidoc" ||
            options.outputFormat == "java-asciidoc") {
            val head = ContentParagraph()
            val descr = ContentParagraph()
            descr.append(ContentText(kdocText))

            val asciidoc = Content.of(head, descr)
            return asciidoc to { node -> };
        }
```

I really hope that you can accept or will accept, after fixing your comments, so we can switch to port our huge java codebase to kotlin.

Cheers,
 Mario

### CODE REVIEW

**Comment by mariotoffia** [comment]
@yole Hi, It's me again... I wonder if you could spend a little while checking my pull request?

Cheers,
 Mario :)

**Comment by mariotoffia** [comment]
Fixed a merge error in `DocumentationBuilder` so now it is clean except for the _.idea/*_ "junk/xml" that I don't know how to handle (since they are not part of the _.gitignore_).

**Comment by mariotoffia** [comment]
@yole @semoro Any comments?

Cheers,
 Mario

File: .idea/.name [hunk 1]
```diff
@@ -1 +0,0 @@
-dokka
```
Comment (semoro) [inline .idea/.name]: Revert this

File: .idea/artifacts/dokka_jar.xml [hunk 2]
```diff
@@ -1,9 +0,0 @@
-<component name="ArtifactManager">
```
Comment (semoro) [inline .idea/artifacts/dokka_jar.xml]: Revert this

File: .idea/artifacts/dokka_zip.xml [hunk 3]
```diff
@@ -1,27 +0,0 @@
-<component name="ArtifactManager">
```
Comment (semoro) [inline .idea/artifacts/dokka_zip.xml]: Revert this

File: .idea/artifacts/javadoc_jar.xml [hunk 4]
```diff
@@ -1,8 +0,0 @@
-<component name="ArtifactManager">
```
Comment (semoro) [inline .idea/artifacts/javadoc_jar.xml]: Revert this

File: .idea/compiler.xml [hunk 5]
```diff
@@ -1,29 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
   <component name="CompilerConfiguration">
-    <wildcardResourcePatterns>
```
Comment (semoro) [inline .idea/compiler.xml]: Revert this

File: .idea/copyright/profiles_settings.xml [hunk 6]
```diff
@@ -1,4 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (semoro) [inline .idea/copyright/profiles_settings.xml]: Revert this

File: .idea/encodings.xml [hunk 7]
```diff
@@ -1,6 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (semoro) [inline .idea/encodings.xml]: Revert this

File: .idea/inspectionProfiles/Project_Default.xml [hunk 8]
```diff
@@ -1,12 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (semoro) [inline .idea/inspectionProfiles/Project_Default.xml]: Revert this

File: .idea/inspectionProfiles/profiles_settings.xml [hunk 9]
```diff
@@ -1,8 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (semoro) [inline .idea/inspectionProfiles/profiles_settings.xml]: Revert this

File: .idea/kotlinc.xml [hunk 10]
```diff
@@ -1,7 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (semoro) [inline .idea/kotlinc.xml]: Revert this

File: .idea/misc.xml [hunk 11]
```diff
@@ -1,29 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
-  <component name="ASMPluginConfiguration">
```
Comment (semoro) [inline .idea/misc.xml]: Revert this

File: .idea/modules.xml [hunk 12]
```diff
@@ -17,7 +17,7 @@
       <module fileurl="file://$PROJECT_DIR$/.idea/modules/core/core.iml" filepath="$PROJECT_DIR$/.idea/modules/core/core.iml" group="core" />
       <module fileurl="file://$PROJECT_DIR$/.idea/modules/core/core_main.iml" filepath="$PROJECT_DIR$/.idea/modules/core/core_main.iml" group="core" />
       <module fileurl="file://$PROJECT_DIR$/.idea/modules/core/core_test.iml" filepath="$PROJECT_DIR$/.idea/modules/core/core_test.iml" group="core" />
-      <module fileurl="file://$PROJECT_DIR$/.idea/modules/dokka.iml" filepath="$PROJECT_DIR$/.idea/modules/dokka.iml" />
```
Comment (semoro) [inline .idea/modules.xml]: Revert this

File: .idea/runConfigurations/Dokka.xml [hunk 13]
```diff
@@ -1,23 +0,0 @@
-<component name="ProjectRunConfigurationManager">
```
Comment (semoro) [inline .idea/runConfigurations/Dokka.xml]: Revert this

File: .idea/runConfigurations/Stdlib.xml [hunk 14]
```diff
@@ -1,23 +0,0 @@
-<component name="ProjectRunConfigurationManager">
```
Comment (semoro) [inline .idea/runConfigurations/Stdlib.xml]: Revert this

File: .idea/runConfigurations/Stdlib_Javadoc.xml [hunk 15]
```diff
@@ -1,15 +0,0 @@
-<component name="ProjectRunConfigurationManager">
```
Comment (semoro) [inline .idea/runConfigurations/Stdlib_Javadoc.xml]: Revert this

File: .idea/scopes/scope_settings.xml [hunk 16]
```diff
@@ -1,5 +0,0 @@
-<component name="DependencyValidationManager">
```
Comment (semoro) [inline .idea/scopes/scope_settings.xml]: Revert this

**Comment by semoro** [comment]
NVM, I will fix merge conflicts before

**Comment by mariotoffia** [comment]
@semoro  Thanks!! - Is there something I can do?

**Comment by semoro** [comment]
Wait, till I will push properly rebased version to your branch, also, please check that `Allow edit from maintainers` are set in PR.
When you checkout project and opened it as IDEA project, you should press import Gradle project in IDEA, and point it to root build.gradle of the project, no further changes in xml will be made if project imported clearly.

**Comment by semoro** [comment]
@mariotoffia Tests in `AsciidocFormatTest` failed
```
kotlin.UninitializedPropertyAccessException: lateinit property docOptions has not been initialized

	at Formats.AsciidocFormatService.getDocOptions(AsciidocFormatService.kt:227)
	at Formats.AsciidocFormatService$formatterOptions$2.invoke(AsciidocFormatService.kt:230)
	at Formats.AsciidocFormatService$formatterOptions$2.invoke(AsciidocFormatService.kt:221)
	at kotlin.SynchronizedLazyImpl.getValue(Lazy.kt:130)
	at Formats.AsciidocFormatService.getFormatterOptions(AsciidocFormatService.kt)
	at Formats.AsciidocFormatService$asciidoctor$2.invoke(AsciidocFormatService.kt:238)
	at Formats.AsciidocFormatService$asciidoctor$2.invoke(AsciidocFormatService.kt:221)
	at kotlin.SynchronizedLazyImpl.getValue(Lazy.kt:130)
	at Formats.AsciidocFormatService.getAsciidoctor(AsciidocFormatService.kt)
	at Formats.AsciidocFormatService.createOutputBuilder(AsciidocFormatService.kt:253)
	at format.AsciidocFormatTest.generate(AsciidocFormatTest.kt:61)
	at format.AsciidocFormatTest.asciiDocTableGetsRendered(AsciidocFormatTest.kt:49)
```

And in `JavaAsciidocTest`

```
Renewing package-list from http://docs.oracle.com/javase/6/docs/api/package-list
  package <root>: 1 declarations
generation completed with 440 warnings
Standard Doclet version 1.8.0_131
Building tree for all the packages and classes...
Generating /tmp/test-4901682236612360102/AsciidocAttributesKt.html...
ERROR: (LoadError) no such file to load -- kalle-gem
java.io.FileNotFoundException: /tmp/test-4901682236612360102/AsciidocAttributesKt.html (No such file or directory)

	at java.io.FileInputStream.open0(Native Method)
	at java.io.FileInputStream.open(FileInputStream.java:195)
	at java.io.FileInputStream.<init>(FileInputStream.java:138)
	at kotlin.io.FilesKt__FileReadWriteKt.readBytes(FileReadWrite.kt:52)
	at kotlin.io.FilesKt__FileReadWriteKt.readText(FileReadWrite.kt:90)
	at kotlin.io.FilesKt__FileReadWriteKt.readText$default(FileReadWrite.kt:90)
	at javadoc.JavaAsciidocTest.generate(JavaAsciidocTest.kt:95)
	at javadoc.JavaAsciidocTest.javaAsciidocInvalidRequireShallNotWork(JavaAsciidocTest.kt:62)
```

**Comment by semoro** [comment]
Anyway I will push your properly rebased changes in our repository, then you can force reset your master to it

**Comment by semoro** [comment]
That's it https://github.com/Kotlin/dokka/tree/PR-183

**Comment by semoro** [comment]
You can check failed tests here
https://teamcity.jetbrains.com/viewLog.html?buildId=1107379&tab=buildResultsDiv&buildTypeId=Kotlin_Dokka_DokkaAntMavenGradle

**Comment by mariotoffia** [comment]
@semoro I've done a:
* git reset --hard upstream/PR-183
* git push origin master --force

Fixed the failing tests.

Strange that my behavior on ```javaAsciidocInvalidRequireShallNotWork``` test did not throw any exception for me (on my _Windows_ box). However did a ```try ... catch``` in order to make the test pass (since it is a negative test).

Committed the changes (excluding _.idea/<xml files>_ that the ide fiddles with) so it should be clean to merge.

Thanks for all help! and please comment if there's something I need to do

Cheers,
 Mario

**Comment by semoro** [comment]
I will review your code later today or yesterday and will write more comments

**Comment by semoro** [comment]
(Will continue tomorrow... So busy today)

**Comment by mariotoffia** [comment]
@semoro Thanks! I've fixed the tabs.

However I did not find any _code style settings xml_ for the project. I'm new to IDEA - is there a another way of defining the code style that I shall use?

My current approach was to set the tabs and indent to 4 and hit CTRL+ALT+SHIFT+L in each file I formatted.

Cheers,
 Mario

**Comment by mariotoffia** [comment]
@semoro Did you have a chance to look at my code?

Cheers,
 Mario

File: core/src/test/kotlin/format/AsciidocFormatTest.kt [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+import kotlin.test.assertEquals
+
+/**
+ * @author Mario Toffia
+ */
+class AsciidocFormatTest {
+    @Test fun asciidocOptionsGetParsedCorrectly() {
+        var options = """
+								--base-dir the-base-dir
+								--attribute
```
Comment (semoro) [inline core/src/test/kotlin/format/AsciidocFormatTest.kt]: val

File: core/src/test/kotlin/format/AsciidocFormatTest.kt [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+import org.jetbrains.dokka.tests.tempLocation
+import org.junit.Test
+import kotlin.test.assertEquals
+
+/**
+ * @author Mario Toffia
+ */
+class AsciidocFormatTest {
+    @Test fun asciidocOptionsGetParsedCorrectly() {
+        var options = """
```
Comment (semoro) [inline core/src/test/kotlin/format/AsciidocFormatTest.kt]: val

File: core/src/test/kotlin/format/AsciidocFormatTest.kt [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+import org.jetbrains.dokka.tests.appendDocumentation
+import org.jetbrains.dokka.tests.tempLocation
+import org.junit.Test
+import kotlin.test.assertEquals
+
+/**
+ * @author Mario Toffia
+ */
+class AsciidocFormatTest {
+    @Test fun asciidocOptionsGetParsedCorrectly() {
```
Comment (semoro) [inline core/src/test/kotlin/format/AsciidocFormatTest.kt]: Please move to separate test

File: core/src/test/kotlin/format/AsciidocFormatTest.kt [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+    private fun generate(file: String): String {
+        val sb = StringBuilder()
+        val cr = contentRootFromPath(file)
+        val documentation = DocumentationModule("test")
+
+        val options = DocumentationOptions("", "html-as-asciidoc",
+                                           includeNonPublic = true,
+                                           skipEmptyPackages = false,
+                                           includeRootPackage = true,
+                                           sourceLinks = listOf<DokkaConfiguration.SourceLinkDefinition>(),
```
Comment (semoro) [inline core/src/test/kotlin/format/AsciidocFormatTest.kt]: Redundant type-parameter

File: core/src/test/kotlin/format/AsciidocFormatTest.kt [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+                                           generateIndexPages = false,
+                                           noStdlibLink = true,
+                                           cacheRoot = "default")
+
+        appendDocumentation(documentation, cr,
+                            withJdk = false,
+                            withKotlinRuntime = false,
+                            options = options)
+
+        val memb = documentation.members.single().members.single().members
```
Comment (semoro) [inline core/src/test/kotlin/format/AsciidocFormatTest.kt]: `members`

File: core/src/test/kotlin/format/AsciidocFormatTest.kt [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+        documentation.prepareForGeneration(options)
+
+        val outputInjector = Guice.createInjector(DokkaOutputModule(options, DokkaConsoleLogger))
+        outputInjector.getInstance(FormatService::class.java).createOutputBuilder(sb, tempLocation).appendNodes(memb);
```
Comment (semoro) [inline core/src/test/kotlin/format/AsciidocFormatTest.kt]: Redundant semi-colon here :slightly_smiling_face:

File: core/src/test/kotlin/format/AsciidocFormatTest.kt [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @Test fun asciiDocTableGetsRendered() {
+        val str = generate("testdata/format/asciidocTable.kt")
+        System.out.println(str)
+    }
+
+    @Test fun asciiDocJavaSourceGetsRendered() {
+        val str = generate("testdata/format/asciidocJavaSource.kt")
+        System.out.println(str)
```
Comment (semoro) [inline core/src/test/kotlin/format/AsciidocFormatTest.kt]: Why tests just print generated text? 
Please add assertion to check that generated text is correct (See how other format tests work)

File: core/src/test/kotlin/javadoc/JavaAsciidocTest.kt [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+
+/**
+ * @author Mario Toffia
+ */
+class JavaAsciidocTest {
+    @Test fun javaAsciidocJavaSourceGetsRendered() {
+        val str = generate("testdata/format/asciidocJavaSource.kt", "AsciidocJavaSourceKt.html")
+
+        assertTrue {
+            str.contains("<pre class=\"CodeRay highlight\"><code data-lang=\"java\"><span " +
```
Comment (semoro) [inline core/src/test/kotlin/javadoc/JavaAsciidocTest.kt]: It's better to compare with a local file, see how other format tests work.
And for other tests here.

File: core/src/test/kotlin/javadoc/JavaAsciidocTest.kt [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+        assertTrue { str.contains("<col style=\"width: 50%;\">") }
+        assertTrue {
+            str.contains("<td class=\"tableblock halign-left valign-top\"><p " +
+                         "class=\"tableblock\">Firefox</p></td>")
+        }
+    }
+
+    @Test fun javaAsciidocLinksIsRenderedCorrectly() {
+        val str = generate("testdata/format/AsciidocJavaLinks.java", "AsciidocJavaLinks.html")
+        println(str)
```
Comment (semoro) [inline core/src/test/kotlin/javadoc/JavaAsciidocTest.kt]: Printing again :slightly_smiling_face:

File: core/src/test/kotlin/javadoc/JavaAsciidocTest.kt [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+    private fun generate(file: String, htmlFile: String, additionalParams: String? = null): String {
+        val cr = contentRootFromPath(file)
+        val documentation = DocumentationModule("test")
+        val tmpDir = Files.createTempDirectory("test-")
+
+        val options = DocumentationOptions(tmpDir.toString(), "java-asciidoc",
+                                           includeNonPublic = true,
+                                           skipEmptyPackages = false,
+                                           includeRootPackage = true,
+                                           sourceLinks = listOf<DokkaConfiguration.SourceLinkDefinition>(),
```
Comment (semoro) [inline core/src/test/kotlin/javadoc/JavaAsciidocTest.kt]: Redundant type parameter

File: core/src/main/kotlin/javadoc/tags.kt [hunk 27]
```diff
@@ -130,36 +153,37 @@ class ThrowsTagAdapter(val holder: Doc, val type: ClassDocumentationNodeAdapter,
...
     override fun position(): SourcePosition? = holder.position()
 
     override fun text(): String = "@return $content" // Seems has no effect, so used for debug
-    override fun inlineTags(): Array<Tag> = buildInlineTags(module, holder, content).toTypedArray()
+    override fun inlineTags(): Array<Tag> = buildInlineTags(module, holder, content, formatService).toTypedArray()
     override fun firstSentenceTags(): Array<Tag> = inlineTags()
 }
 
-fun buildInlineTags(module: ModuleNodeAdapter, holder: Doc, tags: List<ContentNode>): List<Tag> = ArrayList<Tag>().apply { tags.forEach { buildInlineTags(module, holder, it, this) } }
+fun buildInlineTags(module: ModuleNodeAdapter, holder: Doc, tags: List<ContentNode>, formatService: StructuredFormatService?): List<Tag> = ArrayList<Tag>().apply { tags.forEach { buildInlineTags(module, holder, it, this, formatService) } }
```
Comment (semoro) [inline core/src/main/kotlin/javadoc/tags.kt]: It possible to store formatService in ModuleNodeAdapter, then you won't need to bypass it all around.

File: core/src/main/kotlin/Asciidoctor/AsciidocDescriptorDocumentationParser.kt [hunk 28]
```diff
@@ -0,0 +1,148 @@
+package org.jetbrains.dokka.asciidoctor
+
+/**
```
Comment (semoro) [inline core/src/main/kotlin/Asciidoctor/AsciidocDescriptorDocumentationParser.kt]: Please move KDoc to just over class, not an import list

File: core/src/main/kotlin/Asciidoctor/AsciidocDescriptorDocumentationParser.kt [hunk 29]
```diff
@@ -0,0 +1,148 @@
...
+import org.jetbrains.kotlin.resolve.DescriptorUtils
+import org.jetbrains.kotlin.resolve.annotations.argumentValue
+import org.jetbrains.kotlin.resolve.constants.StringValue
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+import org.jetbrains.kotlin.resolve.scopes.DescriptorKindFilter
+import org.jetbrains.kotlin.resolve.scopes.getDescriptorsFiltered
+import org.jetbrains.kotlin.resolve.source.PsiSourceElement
+
+class AsciidocDescriptorDocumentationParser
+@Inject constructor(val options: DocumentationOptions,
```
Comment (semoro) [inline core/src/main/kotlin/Asciidoctor/AsciidocDescriptorDocumentationParser.kt]: Why an `@Inject` here if this class not injected? It's better to create an interface for `DescriptorDocumentationParser` and add injection for it, dependent on format descriptor.

File: core/src/main/kotlin/Asciidoctor/AsciidocDescriptorDocumentationParser.kt [hunk 30]
```diff
@@ -0,0 +1,148 @@
...
+                    node.append(deprecationNode, RefKind.Deprecation)
+                }
+            }
+        }
+        return Content.Empty to { node -> }
+    }
+
+    fun KDocSection.getTags(): Array<KDocTag> = PsiTreeUtil.getChildrenOfType(this, KDocTag::class.java) ?: arrayOf()
+
+    private fun MutableContent.addTagToSeeAlso(descriptor: DeclarationDescriptor, seeTag: KDocTag) {
```
Comment (semoro) [inline core/src/main/kotlin/Asciidoctor/AsciidocDescriptorDocumentationParser.kt]: Unused?

File: core/src/main/kotlin/Asciidoctor/AsciidocDescriptorDocumentationParser.kt [hunk 31]
```diff
@@ -0,0 +1,148 @@
...
+                parseResult.deprecatedContent?.let {
+                    val deprecationNode = DocumentationNode("", it, NodeKind.Modifier)
+                    node.append(deprecationNode, RefKind.Deprecation)
+                }
+            }
+        }
+        return Content.Empty to { node -> }
+    }
+
+    fun KDocSection.getTags(): Array<KDocTag> = PsiTreeUtil.getChildrenOfType(this, KDocTag::class.java) ?: arrayOf()
```
Comment (semoro) [inline core/src/main/kotlin/Asciidoctor/AsciidocDescriptorDocumentationParser.kt]: Unused?

File: core/src/main/kotlin/Asciidoctor/AsciidocDescriptorDocumentationParser.kt [hunk 32]
```diff
@@ -0,0 +1,148 @@
...
+                else                   -> "[$content]"
+            }
+        }
+
+        val head = ContentParagraph()
+        val descr = ContentParagraph()
+        descr.append(ContentText(res))
+
+        val asciidoc = Content.of(head, descr)
+        return asciidoc to { node -> };
```
Comment (semoro) [inline core/src/main/kotlin/Asciidoctor/AsciidocDescriptorDocumentationParser.kt]: Redundant semi-colon :slightly_smiling_face:

File: core/src/main/kotlin/Formats/AsciidocFormatService.kt [hunk 33]
```diff
@@ -0,0 +1,254 @@
...
+    val formatterOptions: FormatterOptions by lazy {
+        parseOptions(docOptions.additionalParams ?: "")
+    }
+
+    val options: Options by lazy {
+        defaultOptions().attributes(defaultAttributes().addFromOptions(formatterOptions)).get()
+    }
+
+    val asciidoctor: Asciidoctor by lazy {
+        var ad = Asciidoctor.Factory.create(formatterOptions.singleOption("gem-path"))
```
Comment (semoro) [inline core/src/main/kotlin/Formats/AsciidocFormatService.kt]: val?

File: core/src/main/kotlin/Formats/AsciidocFormatService.kt [hunk 34]
```diff
@@ -0,0 +1,254 @@
...
+            is ContentEntity     -> sb.append(content.text)
+            else                 -> {
+                flushBuffer()
+                super.appendContent(content)
+            }
+        }
+    }
+
+    private fun flushBuffer() {
+        if (sb.length > 0) {
```
Comment (semoro) [inline core/src/main/kotlin/Formats/AsciidocFormatService.kt]: `isNotEmpty`

File: core/src/main/kotlin/javadoc/tags.kt [hunk 35]
```diff
@@ -4,8 +4,27 @@ import com.sun.javadoc.*
...
 
-class TagImpl(val holder: Doc, val name: String, val text: String): Tag {
-    override fun text(): String? = text
+private object NoopLocation : Location {
+    override val path: String
+        get() = ""
+
+    override fun relativePathTo(other: Location, anchor: String?): String = ""
+}
+fun String.serviceFormat(formatService: StructuredFormatService?) : String? {
```
Comment (semoro) [inline core/src/main/kotlin/javadoc/tags.kt]: Maybe `formatWithStructuredFormatService` is better?

File: core/src/test/kotlin/format/AsciidocFormatTest.kt (hunk 21 above)
Comment (mariotoffia) [inline core/src/test/kotlin/format/AsciidocFormatTest.kt]: @semoro If I do ```documentation.members``` I don't get the correct output. Is there something else I should do?

File: core/src/test/kotlin/javadoc/JavaAsciidocTest.kt (hunk 24 above)
Comment (mariotoffia) [inline core/src/test/kotlin/javadoc/JavaAsciidocTest.kt]: @semoro Is it possible for I can get the inline HTML in this pull request. I will have one more in august where I'll add more unit test and I put the HTML in separate files. Is this OK? If no I'll fix it :)

File: core/src/main/kotlin/Asciidoctor/AsciidocDescriptorDocumentationParser.kt (hunk 28 above)
Comment (mariotoffia) [inline core/src/main/kotlin/Asciidoctor/AsciidocDescriptorDocumentationParser.kt]: @semoro I'm unsure what you mean here?

File: core/src/main/kotlin/Asciidoctor/AsciidocDescriptorDocumentationParser.kt (hunk 29 above)
Comment (mariotoffia) [inline core/src/main/kotlin/Asciidoctor/AsciidocDescriptorDocumentationParser.kt]: @semoro  I didn't want to change such a core interface in order support the asciidoc. But I can - should I?

**Comment by mariotoffia** [comment]
@semoro Hi and thanks for your comments! I really appreciate your time! :) I've fixed all but the description parser and put HTML in separate files.
1) Can I submit a new pull request regarding the external HTML files in agust along with more fixes on asciidoc implementation?
2) I did no want to change such a core feature of dokka as the format service interface. Should I do that? Then I'll do it :)

Cheers,
 Mario

**Comment by mariotoffia** [comment]
@yole That's completely true, since I do not wish to re-implement asciidoctor. If you guys do not like this just delete this pull request and we'll use it internally in our builds.

Cheers,
 Mario

**Comment by MarcinAman** [comment]
Hi,
since then dokka has been rewritten. If you would like to see this feature in new dokka please adjust the code and make another PR

**Pull request closed without merging**